### PR TITLE
Implement the breadcrumb component in the header

### PR DESCRIPTION
**Pull request by anthonydillon**
## Done
Implement the breadcrumb component in the header
Drive-by, fix error while awaiting modelStatusData to become available

## QA
- Load the demo and go to a model details
- See that the breadcrumb renders just the model name
- Go into an application and check the breadcrumb now matches [the design](https://zpl.io/aN63dv4)
- To test the drive-by, load the branch locally
- Go to an application details
- Refresh and see that the page loads without an error

## Details
Fixes https://github.com/canonical-web-and-design/jaas-dashboard/issues/891

## Screenshots
image

### CODE REVIEW

File: src/app/utils/utils.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ export function generateEntityIdentifier(
   baseAppURL,
   disableLink = false
 ) {
+  if (!namespace) {
+    return [];
```
Comment (hatched) [inline src/app/utils/utils.js:15]: This should return `null`. An array is of a different type than a `JSX.Element` which is what this normally returns. If this was in TS it would have complained ;)

File: src/components/Breadcrumb/Breadcrumb.test.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        <MemoryRouter
+          initialEntries={["/models/eggman@external/group-test/app/easyrsa"]}
+        >
+          <TestRoute path="/models/:userName/:modelName?/app/:appName?">
+            <Breadcrumb />
+          </TestRoute>
+        </MemoryRouter>
+      </Provider>
+    );
+    expect(wrapper.find("[data-test='breadcrumb-model']").text()).toStrictEqual(
```
Comment (hatched) [inline src/components/Breadcrumb/Breadcrumb.test.tsx:44]: I was expecting this test to check for something like "group-test > Applications > easyrsa" instead of grabbing each element individually. Can you move the `find` up to `p-breadcrumbs__items` to achieve this?

File: src/components/Breadcrumb/Breadcrumb.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { Link } from "@canonical/react-components";
+import { useParams } from "react-router-dom";
+
+export type EntityDetailsRoute = {
+  userName: string;
+  modelName: string;
+  appName: string;
+};
+
+const Breadcrumb = (): JSX.Element => {
```
Comment (hatched) [inline src/components/Breadcrumb/Breadcrumb.tsx:10]: Just to maintain consistency, we typically use the `function` signature for components and then export them when they are defined...

```ts
export default function Breadcrumb(): JSX.Element {
```

File: src/components/Breadcrumb/Breadcrumb.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import { Link } from "@canonical/react-components";
+import { useParams } from "react-router-dom";
+
+export type EntityDetailsRoute = {
```
Comment (hatched) [inline src/components/Breadcrumb/Breadcrumb.tsx:4]: You can import this type from the `Routes` component.

```ts
import type { EntityDetailsRoute } from "components/Routes/Routes";
```

File: src/components/Breadcrumb/Breadcrumb.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  return (
+    <nav className="p-breadcrumbs" aria-label="Breadcrumb navigation">
+      <ol className="p-breadcrumbs__items">
+        {appName ? (
+          <>
+            <li className="p-breadcrumbs__item" data-test="breadcrumb-model">
+              <Link
+                href={`/models/${
+                  userName ? `${userName}/${modelName}` : `${modelName}`
```
Comment (hatched) [inline src/components/Breadcrumb/Breadcrumb.tsx:21]: This logic is a bit hard to follow and repeated, can you abstract it out into a utility function in this file?

**Review by hatched — CHANGES_REQUESTED** [review]
Thanks @anthonydillon just a few small cleanups requested

File: src/components/Breadcrumb/Breadcrumb.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  const { userName, modelName, appName } = useParams<EntityDetailsRoute>();
+
+  return (
+    <nav className="p-breadcrumbs" aria-label="Breadcrumb navigation">
+      <ol className="p-breadcrumbs__items">
+        {appName ? (
+          <>
+            <li className="p-breadcrumbs__item" data-test="breadcrumb-model">
+              <Link
+                href={`/models/${
```
Comment (hatched) [inline src/components/Breadcrumb/Breadcrumb.tsx:20]: Sorry i missed this in the original review...this should probably use `to`

https://reactrouter.com/web/api/Link

File: src/app/utils/utils.js (hunk 1 above)
Comment (anthonydillon) [reply src/app/utils/utils.js:15]: Sure, `tableRows.js` does this a lot so thought it was the way you guys do it.

Comment (hatched) [reply src/app/utils/utils.js:15]: Ahh np, so the difference there is that on the 'green' path it will return an array of data. So if there is missing data it'll still return an empty array (the same type of data).

File: src/components/Breadcrumb/Breadcrumb.test.tsx (hunk 2 above)
Comment (anthonydillon) [reply src/components/Breadcrumb/Breadcrumb.test.tsx:44]: Not sure a test to check the order is correct is that valuable but I have added it.

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 3 above)
Comment (anthonydillon) [reply src/components/Breadcrumb/Breadcrumb.tsx:10]: Seems to be a mix in the codebase but no problem

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 4 above)
Comment (anthonydillon) [reply src/components/Breadcrumb/Breadcrumb.tsx:4]: Nice, done

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 5 above)
Comment (anthonydillon) [reply src/components/Breadcrumb/Breadcrumb.tsx:21]: Good idea

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 6 above)
Comment (anthonydillon) [reply src/components/Breadcrumb/Breadcrumb.tsx:20]: Thanks done

**Comment by anthonydillon** [comment]
@hatched thanks, think I have addressed all your comments.

File: src/components/Breadcrumb/Breadcrumb.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  return (
+    <nav className="p-breadcrumbs" aria-label="Breadcrumb navigation">
+      <ol className="p-breadcrumbs__items" data-test="breadcrumb-items">
+        {appName ? (
+          <>
+            <li className="p-breadcrumbs__item" data-test="breadcrumb-model">
+              <Link to={generateModelURL()}>{modelName}</Link>
+            </li>
+            <li className="p-breadcrumbs__item" data-test="breadcrumb-section">
+              <Link to={generateModelURL()}>Applications</Link>
```
Comment (barrymcgee) [inline src/components/Breadcrumb/Breadcrumb.tsx:26]: I find it strange that the only two clickable breadcrumbs both link to the same place?

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 3 above)
Comment (hatched) [reply src/components/Breadcrumb/Breadcrumb.tsx]: Hmm yeah we should fix that.

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 7 above)
Comment (hatched) [reply src/components/Breadcrumb/Breadcrumb.tsx:26]: This is per design but it'll change as more pages are implemented into the breadcrumb.

**Review by hatched — APPROVED** [review]

**Comment by hatched** [comment]
QA Note:
When going from an application page to the model page the sidebar bounces open and then gets caught under the cursor. I believe this is resolved in Barry's upcoming branch.

File: src/components/Breadcrumb/Breadcrumb.tsx (hunk 7 above)
Comment (anthonydillon) [reply src/components/Breadcrumb/Breadcrumb.tsx:26]: I agree, and design may change there mind now its in practise.

**Comment by ziheliu214** [comment]
When I refresh the page it shows: 
Screenshot 2021-03-03 at 11 25 31

Another issue might need to be addressed separately - every time jumping pages, the left nav has the animation effect expanding then collapsing, a bit hindering the view.

Apart from that, UX is good!

**Comment by barrymcgee** [comment]
@ziheliu214 Refreshing on the demo service will always produce that response, unfortunately.

I have a fix for the bouncing sidebar incoming in a separate branch.

**Comment by ziheliu214** [comment]
> @ziheliu214 Refreshing on the demo service will always produce that response, unfortunately.
> 
> I have a fix for the bouncing sidebar incoming in a separate branch.

Thanks Barry :D
